src/api/procedures/createPortfolios.ts

```diff
@@ -2,71 +2,80 @@ import { ISubmittableResult } from '@polkadot/types/types';
 
 import { Context, NumberedPortfolio, PolymeshError, Procedure } from '~/internal';
 import { ErrorCode, TxTags } from '~/types';
-import { BatchTransactionSpec, ExtrinsicParams } from '~/types/internal';
+import { BatchTransactionSpec, ProcedureAuthorization } from '~/types/internal';
 import {
   identityIdToString,
   stringToBytes,
   stringToIdentityId,
   u64ToBigNumber,
 } from '~/utils/conversion';
-import { filterEventRecords, getPortfolioIdsByName } from '~/utils/internal';
-
+import { checkTxType, filterEventRecords, getPortfolioIdsByName } from '~/utils/internal';
+type PortfolioParams = {
+  name: string;
+  ownerDid?: string;
+};
 /**
  * @hidden
  */
 export interface Params {
-  names: string[];
+  portfolios: PortfolioParams[];
 }
-
 /**
  * @hidden
  */
 export const createPortfoliosResolver =
   (context: Context) =>
   (receipt: ISubmittableResult): NumberedPortfolio[] => {
     const records = filterEventRecords(receipt, 'portfolio', 'PortfolioCreated');
-
     return records.map(({ data }) => {
       const did = identityIdToString(data[0]);
       const id = u64ToBigNumber(data[1]);
-
       return new NumberedPortfolio({ did, id }, context);
     });
   };
-
 /**
  * @hidden
  */
 export async function prepareCreatePortfolios(
   this: Procedure<Params, NumberedPortfolio[]>,
   args: Params
-): Promise<
-  BatchTransactionSpec<NumberedPortfolio[], ExtrinsicParams<'portfolio', 'createPortfolio'>[]>
-> {
+): Promise<BatchTransactionSpec<NumberedPortfolio[], unknown[][]>> {
   const {
     context: {
       polymeshApi: { tx },
     },
     context,
   } = this;
-  const { names: portfolioNames } = args;
-
+  const { portfolios: portfoliosToCreate } = args;
   const { did } = await context.getSigningIdentity();
-
-  const rawIdentityId = stringToIdentityId(did, context);
-
-  const rawNames = portfolioNames.map(name => stringToBytes(name, context));
-
-  const portfolioIds = await getPortfolioIdsByName(rawIdentityId, rawNames, context);
-
+  const groupedPortfolioNamesByIdentity = portfoliosToCreate.reduce(
+    (acc: Record<string, string[]>, { name, ownerDid }) => {
+      const owner = ownerDid ?? did;
+      if (!acc[owner]) {
+        acc[owner] = [];
+      }
+      acc[owner].push(name);
+      return acc;
+    },
+    {}
+  );
+  const portfolioIdCalls = Object.keys(groupedPortfolioNamesByIdentity).map(identityId => {
+    const rawNames = groupedPortfolioNamesByIdentity[identityId].map(name =>
+      stringToBytes(name, context)
+    );
+    const rawIdentityId = stringToIdentityId(identityId, context);
+    return getPortfolioIdsByName(rawIdentityId, rawNames, context);
+  });
+  const dids = Object.keys(groupedPortfolioNamesByIdentity);
```

Comment:
```suggestion
  const dids = Object.keys(groupedPortfolioNamesByIdentity);
  const portfolioIdCalls = dids.map(identityId => {
     ...
  });
```

(nit) Move `dids` to line above, and then use that variable instead of two `Object.keys` calls.